packages/repository/src/model.ts

```diff
@@ -21,11 +21,11 @@ export type PropertyType = string | Function | Object | Type<any>;
  * Property definition for a model
  */
 export interface PropertyDefinition {
-  type: PropertyType; // For example, 'string', String, or {}
+  type: PropertyType | TypeResolver; // For example, 'string', String, or {}
```

Comment:
As I commented above (see https://github.com/strongloop/loopback-next/pull/1618#discussion_r212284888 and https://github.com/strongloop/loopback-next/pull/1618#discussion_r212954581), I'd prefer to design these changes in such way that we don't need `isTypeResolver` at all, because I am concerned about edge cases where it may not be possible to automatically distinguish between a type and a type resolver, especially when classes and function values in general get involved.

How about using a different property to store the type resolver value? The idea is that `PropertyDefinition` should provide either a `type: PropertyType` or a `typeResolver: TypeResolver`.

```ts
export interface PropertyTypeValue {
  type: PropertyType;
}

export interface PropertyResolvedType {
  typeResolver: TypeResolver;
}

export interface PropertyArrayValue {
  type: 'array';
  itemType: PropertyType;
}

export interface PropertyResolvedArray {
  type: 'array';
  itemTypeResolver: PropertyTypeResolver;
}

export interface PropertyTypeDefinition =
 | PropertyTypeValue
 | PropertyResolvedType
 | PropertyArrayValue
 | PropertyResolvedArray;

export interface PropertyDefinition =  PropertyTypeDefinition & {
  id?: boolean;
  // etc.
}
```

This can be accompanied by property decorators accepting a type resolver, for example `@resolvedProperty` and `@resolvedProperty.array`.

Thoughts?

cc @raymondfeng @strongloop/sq-lb-apex

Reply:
An alternative I've also thought of is to introduce a property called `isTypeResolver`, which would be set when the relation decorators and the example decorators you've mentioned above. The pro would be that we wouldn't need to introduce different property type definitions while the con would be that we may need to check whether it's `type` that's a resolver or `itemType`. Thoughts?

Reply:
I see pros and cons on both sides. If we can reliably detect the resolverFunction, I will vote for reusing the same names. 

To generalize the idea, we should try to allow OpenAPI like schema definitions and introduce a separate decorator such as `@schema` to decouple the typing from `@property`?

```ts
@property({name: 'my-prop'})
@schema({type: ...})
myProp: string;
```

The `@schema` should allow one of the following:

1. `{type: 'number'}`
2. `{$ref: '#/components/schemas/Customer'}`
3. `{type: Customer}`
4. `{type: () => Customer}`. (Maybe `{$ref: () => Customer}` is even better)

Reply:
I'm thinking that we don't need `@schema` as long as we have one of the two changes me and miroslav had mentioned implemented.

The only places where a model and its resolver would need to be differentiated is in the JSON schema conversion logic (where a resolver would have to be resolved first) and in `keyTo` inference for `resolveHasManyMetadata`/`resolveBelongsToMetadata`, where the latter option may not even be a concern since we're heading towards the direction of dropping the use case for `@hasMany(AModelConstructor)`.

For the sake of landing this PR, it might be better to stay with the current coupling of providing type information and the juggler.

This reminds me, with the proposed change in this comment thread, do you think that the use case for the relation decorators taking in the model constructor can stay?